File: interview/02.05.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-08-03 19:46:06
LastEditors: lmio 2091319361@qq.com
Description: 面试题 02.05. 链表求和
'''

from utils.node import ListNode


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode()
        node = head
        carry = 0
        while l1 or l2:
            num1, num2 = 0, 0
            if l1:
                num1 = l1.val
                l1 = l1.next
            if l2:
                num2 = l2.val
                l2 = l2.next
            carry, num = divmod(num1 + num2 + carry, 10)
            node.next = ListNode(num)
            node = node.next
        if carry:
            node.next = ListNode(carry)
        return head.next

File: interview/02.05.go
/*
 * @Author: lmio
 * @Date: 2023-03-28 23:32:28
 * @LastEditTime: 2023-08-03 20:50:15
 * @FilePath: /leetcode/interview/02.05.go
 * @Description:面试题 02.05. 链表求和
 */
package interview

import . "leetcode/code_struct"

func AddTwoNumbers(l1, l2 *ListNode) *ListNode {
	carry := 0
	head := &ListNode{}
	node := head
	for l1 != nil || l2 != nil {
		num1, num2 := 0, 0
		if l1 != nil {
			num1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			num2 = l2.Val
			l2 = l2.Next
		}
		sum := num1 + num2 + carry
		node.Next = &ListNode{Val: sum % 10}
		carry = sum / 10
		node = node.Next
	}

	if carry != 0 {
		node.Next = &ListNode{Val: carry}
	}
	return head.Next
}
